pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,471 @@
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type OrderId: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Debug
+            + MaybeDisplay
+            + AtLeast32BitUnsigned
+            + Copy
+            + Ord
+            + PartialEq
+            + Eq
+            + MaxEncodedLen
+            + scale_info::TypeInfo;
+        type MaxOpenedLimitOrdersForAllOrderBooksPerUser: Get<u32>;
+        type MaxLimitOrdersForPrice: Get<u32>;
+        type WeightInfo: WeightInfo;
+    }
+
+    pub type OrderBookId<T> = TradingPair<<T as assets::Config>::AssetId>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn order_books)]
+    pub type OrderBooks<T: Config> =
+        StorageMap<_, Blake2_128Concat, OrderBookId<T>, OrderBook<T>, OptionQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn limit_orders)]
+    pub type LimitOrders<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        OrderBookId<T>,
+        Blake2_128Concat,
+        T::OrderId,
+        LimitOrder<T>,
+        OptionQuery,
+    >;
+
+    #[pallet::storage]
+    #[pallet::getter(fn bids)]
+    pub type Bids<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        OrderBookId<T>,
+        Blake2_128Concat,
+        T::Balance,
+        BoundedVec<T::OrderId, T::MaxLimitOrdersForPrice>,
+        OptionQuery,
+    >;
+
+    #[pallet::storage]
+    #[pallet::getter(fn asks)]
+    pub type Asks<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        OrderBookId<T>,
+        Blake2_128Concat,
+        T::Balance,
+        BoundedVec<T::OrderId, T::MaxLimitOrdersForPrice>,
+        OptionQuery,
+    >;
+
+    #[pallet::storage]
+    #[pallet::getter(fn user_limit_orders)]
+    pub type UserLimitOrders<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Blake2_128Concat,
+        OrderBookId<T>,
+        BoundedVec<T::OrderId, T::MaxOpenedLimitOrdersForAllOrderBooksPerUser>,
+        OptionQuery,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        OrderPlaced {
+            order_book_id: OrderBookId<T>,
+            order_id: T::OrderId,
+            owner_id: T::AccountId,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Trading pair currently reached its capacity
+        OrderLimitReached,
+        /// Price in given order exceeds allowed limits for the trading pair
+        PriceExceedsLimits,
+        /// Lifespan exceeds defined limits
+        InvalidLifespan,
+        /// Order book does not exist for this trading pair
+        UnknownOrderBook,
+        /// Order book already exists for this trading pair
+        OrderBookAlreadyExists,
+        /// Cannot insert the limit order
+        InsertLimitOrderError,
```

Comment:
Apparently, this error is only returned when the limit of bounded vector is reached. I think it will be better if this is represented in the error somehow. Otherwise this (and `DeleteLimitOrderError`) seems pretty uninformative

Reply:
Will fix Insert Error, but leave DeleteLimitOrderError as is. This error should never happen because we will check that order exists before deleting

Reply:
> we will check that order exists before deleting
So all 12 occurrences of this error are impossible to get returned? I'm just worried that if we miss something it will be pretty hard to debug

Reply:
I think the list of 12 different errors for this case is redundant. It should never happen, if it is it means we have big problem in our code related to data consistency.
Moreover, these errors are returned to the user. They must be informative from the user perspective first of all, secondly - for debug.

Reply:
Added comments